### DISCUSSION TITLE
graphql/codegen/docs/guides/api-testing

### DISCUSSION THREAD

**Discussion by giscus** [dotansimha/graphql-code-generator#10134] [Docs Discussions]
# graphql/codegen/docs/guides/api-testing

GraphQL Code Generator can generate typed results for your GraphQL operations. This tutorial uses GraphQL Yoga, but the practices are applicable to testing any GraphQL server.

https://the-guild.dev/graphql/codegen/docs/guides/api-testing

**Comment by garfieldnate** [comment]
I'm here trying to migrate from Apollo to Yoga, having googled `executeOperation`. This is a method available on `ApolloServer`, and I thought it would be available on `YogaServer`, but it is not. google brought me here, but I can see that `executeOperation` here specifies URLs and likely goes through the entire request stack, which is something I would like to avoid for my use-case.

**Reply by garfieldnate** [reply]
Looks like what I wanted was `YogaServer.getResultForParams`. It requires more parameters and the typing is clumsier than the Apollo usage. Here's my code for reference:

```ts
import { createYoga } from "graphql-yoga";
const {
  buildClientSchema,
  getIntrospectionQuery,
  printSchema
} = require("graphql");

// Either this or import YogaServer and add generic params everywhere it is used, since it has no defaults
type YogaServer = ReturnType<typeof createYoga>;

function isNotAsyncIterator<T extends object>(
  result: T | AsyncIterable<T>,
): result is T {
  return !("next" in result);
}

const writeSchemaToString = async (server: YogaServer) => {
  const result = await server.getResultForParams(
    {
      params: { query: getIntrospectionQuery() },
      request: {} as any,
      batched: false,
    },
    {},
  );
  // narrow type
  assert(result);
  assert(isNotAsyncIterator(result));

  const schema = buildClientSchema(result.data);
  console.log(printSchema(schema));
}
```

**Comment by DanteUkraine** [comment]
I prefer solution that generates type safe client with autogenerated queries and mutations for integration testing: https://www.npmjs.com/package/playwright-graphql
